src/pathTemplate.ts

```diff
@@ -118,69 +99,134 @@ export class PathTemplate {
-      if (segment.kind === extras.BINDING) {
-        if (!has(bindings, segment.literal)) {
-          const msg = util.format(
-            'Value for key %s is not provided in %s',
-            segment.literal,
-            bindings
-          );
-          throw new TypeError(msg);
-        }
-        const tmp = new PathTemplate(bindings[segment.literal].toString());
-        Array.prototype.push.apply(out, tmp.segments);
-        inABinding = true;
-      } else if (segment.kind === extras.END_BINDING) {
-        inABinding = false;
-      } else if (inABinding) {
-        return;
-      } else {
-        out.push(segment);
+    if (Object.keys(bindings).length !== Object.keys(this.bindings).length) {
+      throw new TypeError('The number of variables does not match');
+    }
+    let path = this.inspect();
+    for (const key of Object.keys(bindings)) {
+      const b = bindings[key].toString();
+      if (!this.bindings[key]) {
+        throw new TypeError(`render fails for not matching ${bindings[key]}`);
       }
-    });
+      const variable = this.bindings[key];
 
-    const result = formatSegments(out);
-    this.match(result);
-    return result;
+      if (variable === '*') {
+        if (!b.match(/[^/{}]+/)) {
+          throw new TypeError(`render fails for not matching ${b}`);
+        }
+        path = path.replace(`{${key}=*}`, `${b}`);
+      } else if (variable === '**') {
+        if (!b.match(/[^{}]+/)) {
+          throw new TypeError(`render fails for not matching ${b}`);
+        }
+        path = path.replace(`{${key}=**}`, `${b}`);
+      }
+    }
+    return path;
   }
 
   /**
    * Renders the path template.
    *
    * @return {string} contains const names matched to binding values
    */
-  inspect() {
-    return formatSegments(this.segments);
+  inspect(): string {
+    return this.segments.join('/');
+  }
+  /**
+   * Parse the path template.
+   *
+   * @return {string[]} return segments of the input path.
+   * For example: 'buckets/{hello}'' will give back ['buckets', {hello=*}]
+   */
+  private parsePathTemplate(data: string): string[] {
+    const pathSegments = splitPathTemplate(data);
+    let index = 0;
+    let wildCardCount = 0;
+    const segments: string[] = [];
+    pathSegments.forEach(segment => {
+      // * or ** -> segments.push('{$0=*}');
+      //         -> bindings['$0'] = '*'
+      if (segment === '*' || segment === '**') {
+        this.bindings[`$${index}`] = segment;
+        segments.push(`{$${index}=${segment}}`);
+        index = index + 1;
+        if (segment === '**') {
+          wildCardCount = wildCardCount + 1;
+          if (wildCardCount > 1) {
+            throw new TypeError('Can not have more than one wildcard.');
+          }
+        }
+      }
+      // {project} / {project=*} -> segments.push('{project=*}');
+      //           -> bindings['project'] = '*'
+      else if (segment.match(/(?<={)[0-9a-zA-Z-.~_]+(=\*)?(?=})/)) {
```

Comment:
Look behinds are nice! :) 

I will need some time to review this logic, please don't merge just yet.

Reply:
sounds great! Appreciate that!